src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Runtime.Augments;
+
+namespace System.Threading
+{
+    internal static partial class ClrThreadPool
+    {
+        public static class WaitThread
+        {
+            private struct CompletedWaitHandle
+            {
+                public CompletedWaitHandle(RegisteredWaitHandle completedHandle, bool timedOut)
+                {
+                    CompletedHandle = completedHandle;
+                    TimedOut = timedOut;
+                }
+
+                public RegisteredWaitHandle CompletedHandle { get; }
+                public bool TimedOut { get; }
+            }
+
+            private static readonly RegisteredWaitHandle[] s_registeredWaitHandles = new RegisteredWaitHandle[WaitHandle.MaxWaitHandles];
+
+            private static int s_numActiveWaits = 0;
+
+            private static WaitHandle[] s_waitHandles = new WaitHandle[0];
+
+            private static int s_currentTimeout = Timeout.Infinite;
+
+            private static bool s_waitThreadStarted = false;
+
+            private static LowLevelMonitor s_waitThreadStartedMonitor = new LowLevelMonitor();
+
+            private static LowLevelMonitor s_activeWaitMonitor = new LowLevelMonitor();
+
+            private static void WaitThreadStart()
+            {
+                s_waitThreadStartedMonitor.Acquire();
+                s_waitThreadStarted = true;
+                s_waitThreadStartedMonitor.Signal_Release();
+
+                while (true)
+                {
+                    s_activeWaitMonitor.Acquire();
```

Comment:
I don't think these monitors are necessary. The wait thread can add an event to the list of registered events for interrupting the WaitAny to add/remove waits and it can wait on that event when there are no registered events until a timeout.